hoomd/mpcd/CollisionMethod.cu

```diff
@@ -0,0 +1,263 @@
+// Copyright (c) 2009-2025 The Regents of the University of Michigan.
+// Part of HOOMD-blue, released under the BSD 3-Clause License.
+
+/*!
+ * \file mpcd/CollisionMethod.cu
+ * \brief Defines GPU functions and kernels used by mpcd::CollisionMethod
+ */
+
+#include "CollisionMethod.cuh"
+#include "hoomd/ParticleData.cuh"
+
+namespace hoomd
+    {
+namespace mpcd
+    {
+namespace gpu
+    {
+namespace kernel
+    {
+
+__global__ void store_initial_embedded_group_velocities(Scalar4* d_initial_vel,
+                                                        const Scalar4* d_velocity,
+                                                        const unsigned int* d_embed_group,
+                                                        const unsigned int num_group)
+    {
+    // one thread per particle
+    unsigned int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    if (idx >= num_group)
+        return;
+
+    // store the initial velocity
+    unsigned int particle_idx = d_embed_group[idx];
+    d_initial_vel[idx] = d_velocity[particle_idx];
+    }
+
+__global__ void accumulate_rigid_body_momenta(Scalar3* d_linmom_accum,
+                                              Scalar3* d_angmom_accum,
+                                              const Scalar4* d_initial_vel,
+                                              const unsigned int* d_embed_group,
+                                              const Scalar4* d_postype,
+                                              const Scalar4* d_velocity,
+                                              const int3* d_image,
+                                              const unsigned int* d_body,
+                                              const unsigned int* d_rtag,
+                                              const BoxDim global_box,
```

Comment:
Just wanted to note that here you are correct to pass by value, not by reference, for the reasons we discussed